src/ILCompiler.Compiler/src/Compiler/RyuJitCompilation.cs

```diff
@@ -0,0 +1,141 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+
+using ILCompiler.DependencyAnalysis;
+using ILCompiler.DependencyAnalysisFramework;
+
+using Internal.IL;
+using Internal.TypeSystem;
+using Internal.JitInterface;
+
+using Debug = System.Diagnostics.Debug;
+
+namespace ILCompiler
+{
+    public sealed class RyuJitCompilation : Compilation
+    {
+        private CorInfoImpl _corInfo;
+        private JitConfigProvider _jitConfigProvider;
+
+        internal RyuJitCompilation(
+            DependencyAnalyzerBase<NodeFactory> dependencyGraph,
+            NodeFactory nodeFactory,
+            Logger logger,
+            JitConfigProvider configProvider)
+            : base(dependencyGraph, nodeFactory, new NameMangler(false), logger)
+        {
+            _jitConfigProvider = configProvider;
+        }
+
+        protected override void CompileInternal(string outputFile)
+        {
+            _corInfo = new CorInfoImpl(this, _jitConfigProvider);
+
+            var nodes = _dependencyGraph.MarkedNodeList;
+
+            ObjectWriter.EmitObject(outputFile, nodes, NodeFactory);
+        }
+
+        protected override void ComputeDependencyNodeDependencies(List<DependencyNodeCore<NodeFactory>> obj)
+        {
+            foreach (DependencyNodeCore<NodeFactory> dependency in obj)
+            {
+                var methodCodeNodeNeedingCode = dependency as MethodCodeNode;
+                if (methodCodeNodeNeedingCode == null)
+                {
+                    // To compute dependencies of the shadow method that tracks dictionary
+                    // dependencies we need to ensure there is code for the canonical method body.
+                    var dependencyMethod = (ShadowConcreteMethodNode<MethodCodeNode>)dependency;
+                    methodCodeNodeNeedingCode = dependencyMethod.CanonicalMethodNode;
+                }
+
+                // We might have already compiled this method.
+                if (methodCodeNodeNeedingCode.StaticDependenciesAreComputed)
+                    continue;
+
```

Comment:
Yes, I thought about it (and it's the reason why `ShadowConcreteMethodNode` is generic over whatever is the code node in the compilation) - I specifically avoided this though because this is the place where we're going to make the compiler multi-threaded. That work should not be slowed down by having to multi-thread cppcodegen too.